File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let first_then_apply (lst : 'a list) (pred : 'a -> bool) (f : 'a -> 'b option) : 'b option =
  let rec find_and_apply = function
    | [] -> None
    | x :: xs -> if pred x then f x else find_and_apply xs
  in
  find_and_apply lst

let rec int_pow (base : int) (exponent : int) : int =
  if exponent = 0 then 1 else base * int_pow base (exponent - 1)

let powers_generator (base : int) : int Seq.t =
  let rec aux (n : int) : int Seq.node =
    Seq.Cons (int_pow base n, fun () -> aux (n + 1))
  in
  fun () -> aux 0

let meaningful_line_count (filename : string) : int =
  let count_lines (ic : in_channel) : int =
    let rec aux (count : int) : int =
      try
        let line = input_line ic in
        let trimmed_line = String.trim line in
        if trimmed_line = "" || String.get trimmed_line 0 = '#' then
          aux count
        else
          aux (count + 1)
      with End_of_file -> count
    in
    aux 0
  in
  let ic = open_in filename in
  Fun.protect (fun () -> count_lines ic) ~finally:(fun () -> close_in ic)

type shape =
  | Sphere of float
  | Box of float * float * float

let volume (s : shape) : float =
  match s with
  | Sphere r -> (4.0 /. 3.0) *. Float.pi *. (r ** 3.0)
  | Box (l, w, h) -> l *. w *. h

let surface_area (s : shape) : float =
  match s with
  | Sphere r -> 4.0 *. Float.pi *. (r ** 2.0)
  | Box (l, w, h) -> 2.0 *. (l *. w +. w *. h +. h *. l)

type 'a bst =
  | Empty
  | Node of 'a * 'a bst * 'a bst

let rec insert (x : 'a) (tree : 'a bst) : 'a bst =
  match tree with
  | Empty -> Node (x, Empty, Empty)
  | Node (y, left, right) ->
      if x < y then Node (y, insert x left, right)
      else if x > y then Node (y, left, insert x right)
      else tree

let rec contains (x : 'a) (tree : 'a bst) : bool =
  match tree with
  | Empty -> false
  | Node (y, left, right) ->
      if x = y then true
      else if x < y then contains x left
      else contains x right

let rec size (tree : 'a bst) : int =
  match tree with
  | Empty -> 0
  | Node (_, left, right) -> 1 + size left + size right

let rec inorder (tree : 'a bst) : 'a list =
  match tree with
  | Empty -> []
  | Node (y, left, right) -> inorder left @ [y] @ inorder right

File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(
  arr: T[],
  predicate: (value: T) => boolean,
  fn: (value: T) => U
): U | undefined {
  const found = arr.find(predicate);
  return found !== undefined ? fn(found) : undefined;
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  let power = 0n;
  while (true) {
    yield base ** power;
    power += 1n;
  }
}

export async function meaningfulLineCount(filePath: string): Promise<number> {
  const fileHandle = await open(filePath, 'r');
  let count = 0;

  try {
    for await (const line of fileHandle.readLines()) {
      if (line.trim() !== '' && !line.trim().startsWith('#')) {
        count++;
      }
    }
  } finally {
    await fileHandle.close();
  }

  return count;
}

export type Shape =
  | { kind: "Box"; width: number; length: number; depth: number }
  | { kind: "Sphere"; radius: number };

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return shape.width * shape.length * shape.depth;
    case "Sphere":
      return (4 / 3) * Math.PI * shape.radius ** 3;
  }
}

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return 2 * (shape.width * shape.length + shape.width * shape.depth + shape.length * shape.depth);
    case "Sphere":
      return 4 * Math.PI * shape.radius ** 2;
  }
}

class Node<T> {
  constructor(
    public value: T,
    public left: BinarySearchTree<T> = new Empty(),
    public right: BinarySearchTree<T> = new Empty()
  ) {}
}

export abstract class BinarySearchTree<T> {
  abstract insert(value: T): BinarySearchTree<T>;
  abstract contains(value: T): boolean;
  abstract size(): number;
  abstract inorder(): IterableIterator<T>;
  abstract toString(): string;
}

export class Empty<T> extends BinarySearchTree<T> {
  insert(value: T): BinarySearchTree<T> {
    return new NonEmpty(value);
  }

  contains(): boolean {
    return false;
  }

  size(): number {
    return 0;
  }

  *inorder(): IterableIterator<T> {
  }

  toString(): string {
    return "()";
  }
}

export class NonEmpty<T> extends BinarySearchTree<T> {
  constructor(
    private value: T,
    private left: BinarySearchTree<T> = new Empty(),
    private right: BinarySearchTree<T> = new Empty()
  ) {
    super();
  }

  insert(value: T): BinarySearchTree<T> {
    if (value < this.value) {
      return new NonEmpty(this.value, this.left.insert(value), this.right);
    } else if (value > this.value) {
      return new NonEmpty(this.value, this.left, this.right.insert(value));
    } else {
      return this; 
    }
  }

  contains(value: T): boolean {
    if (value < this.value) {
      return this.left.contains(value);
    } else if (value > this.value) {
      return this.right.contains(value);
    } else {
      return true;
    }
  }

  size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  *inorder(): IterableIterator<T> {
    yield* this.left.inorder();
    yield this.value;
    yield* this.right.inorder();
  }

  toString(): string {
    const leftStr = this.left.toString();
    const rightStr = this.right.toString();
    return `(${leftStr === "()" ? "" : leftStr}${this.value}${rightStr === "()" ? "" : rightStr})`;
  }
}
